iree/compiler/Conversion/LinalgToLLVM/LinalgTileAndVectorizePass.cpp

```diff
@@ -0,0 +1,161 @@
+struct TileWorkgroups : public linalg::LinalgBaseTilingPattern {
+  using Base = linalg::LinalgBaseTilingPattern;
+  TileWorkgroups(MLIRContext *context, linalg::LinalgTilingOptions options,
+                 linalg::LinalgMarker marker, PatternBenefit benefit = 1)
+      : Base(LinalgOpTy::getOperationName(), context, options, marker,
+             benefit) {}
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    // Find the parent FuncOp before tiling. If tiling succeeds, the op will be
+    // erased.
+    FuncOp funcOp = op->getParentOfType<FuncOp>();
+    SmallVector<Value, 4> tensorResults;
+    if (!funcOp ||
+        failed(Base::matchAndRewriteBase(op, rewriter, tensorResults)) ||
+        !tensorResults.empty()) {
+      return failure();
+    }
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+}  // namespace
+
+namespace {
+struct TileAndVectorizeWorkgroups
+    : public PassWrapper<TileAndVectorizeWorkgroups, FunctionPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, AffineDialect, scf::SCFDialect,
+                    vector::VectorDialect>();
+  }
+  void runOnFunction() override;
+};
+}  // namespace
+
+void TileAndVectorizeWorkgroups::runOnFunction() {
+  auto funcOp = getOperation();
+  MLIRContext *context = &getContext();
+
+  OwningRewritePatternList l1patterns, l2patterns;
+  CPUKernelDispatch cpuKernelDispatch;
+
+  // First level of tiling patterns. (workgroups memory)
+  l1patterns.insert<TileWorkgroups<linalg::MatmulOp>,
+                    TileWorkgroups<linalg::BatchMatmulOp>>(
+      context,
+      linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+          [&cpuKernelDispatch](OpBuilder &builder,
+                               Operation *operation) -> SmallVector<Value, 4> {
+            return TileSizeFn::get<TilingLevel::Level1Tiles>(
+                cpuKernelDispatch, builder, operation);
+          }),
+      linalg::LinalgMarker(
+          Identifier::get(getWorkgroupMarker(), context),
+          Identifier::get(getWorkgroupL1TileMarker(), context)));
+
+  // Second level of tiling patterns. (workgroups memroey -> vectors)
+  l2patterns.insert<TileWorkgroups<linalg::MatmulOp>,
+                    TileWorkgroups<linalg::BatchMatmulOp>>(
+      context,
+      linalg::LinalgTilingOptions().setTileSizeComputationFunction(
+          [&cpuKernelDispatch](OpBuilder &builder,
+                               Operation *operation) -> SmallVector<Value, 4> {
+            return TileSizeFn::get<TilingLevel::Level2Tiles>(
+                cpuKernelDispatch, builder, operation);
+          }),
+      linalg::LinalgMarker(Identifier::get(getWorkgroupL1TileMarker(), context),
+                           Identifier::get(getVectorizeMarker(), context)));
+
+  // Apply tiling.
+  applyPatternsAndFoldGreedily(funcOp, std::move(l1patterns));
```

Comment:
Not sure you need the `move`. The function expects a const &.

Reply:
This is needed, The API is update recently using ti use `FrozenRewritePatternList`
https://github.com/llvm/llvm-project/blob/master/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h#L38

Reply:
Ok, I missed that. My issue with such code style is that if you unknowingly use `l1Patterns` later you will get weird behavior.
Two options
1) Use FrozenRewritePatternList here and avoid using move.
2)  My preference is doing something like this
```
{
   OwningRewritePatternList l1PatternList;
   ...
   applyPatternsAndFoldGreedily(..., std::move(l1PatternList))
}
```
This way the list is out of scope when you move it.

Reply:
Done!